source/crypto/signing.cpp

```diff
@@ -162,64 +167,62 @@ signInputs(const Models::Seed& seed, const std::vector<Models::Address>& inputs,
       // Get corresponding private key of address
       auto key = Types::bytesToTrits(Crypto::Signing::key(
-          Types::trytesToBytes(seed.toTrytes()), keyIndex, seed.getSecurity()));  // TODO(optimize)
+          Types::trytesToBytes(seed.toTrytes()), keyIndex, keySecurity));  // TODO(optimize)
 
-      //  First 6561 trits for the firstFragment
+      // First 6561 trits for the firstFragment
       std::vector<int8_t> firstFragment(&key[0], &key[6561]);
 
-      //  Get the normalized bundle hash
+      // Get the normalized bundle hash
       auto normalizedBundleHash = bundle.normalizedBundle(bundleHash);
 
-      //  First bundle fragment uses 27 trytes
+      // First bundle fragment uses 27 trytes
       std::vector<int8_t> firstBundleFragment(&normalizedBundleHash[0], &normalizedBundleHash[27]);
 
-      //  Calculate the new signatureFragment with the first bundle fragment
+      // Calculate the new signatureFragment with the first bundle fragment
       auto firstSignedFragment =
           Crypto::Signing::signatureFragment(firstBundleFragment, firstFragment);
 
-      //  Convert signature to trytes and assign the new signatureFragment
-      tx.setSignatureFragments(Types::tritsToTrytes(firstSignedFragment));
+      // Convert signature to trytes and assign the new signatureFragment
+      bundle[i].setSignatureFragments(Types::tritsToTrytes(firstSignedFragment));
 
       // if user chooses higher than 27-tryte security
       // for each security level, add an additional signature
-      for (int j = 1; j < keySecurity; j++) {
-        //  Because the signature is > 2187 trytes, we need to
-        //  find the second transaction to add the remainder of the signature
-        for (auto& txb : bundle.getTransactions()) {
-          //  Same address as well as value = 0 (as we already spent the input)
-          if (txb.getAddress() == addr && txb.getValue() == 0) {
-            // Use the second 6562 trits
-            std::vector<int8_t> secondFragment(&key[6561], &key[6561 * 2]);
-
-            // The second 27 to 54 trytes of the bundle hash
-            std::vector<int8_t> secondBundleFragment(&normalizedBundleHash[27],
-                                                     &normalizedBundleHash[27 * 2]);
-
-            //  Calculate the new signature
-            auto secondSignedFragment =
-                Crypto::Signing::signatureFragment(secondBundleFragment, secondFragment);
-
-            //  Convert signature to trytes and assign it again to this bundle entry
-            txb.setSignatureFragments(Types::tritsToTrytes(secondSignedFragment));
-          }
+      for (unsigned int j = 1; j < keySecurity; ++j) {
+        // Because the signature is > 2187 trytes, we need to
+        // find the subsequent transaction to add the remainder of the signature
+        // Same address as well as value = 0 (as we already spent the input)
+        if (bundle[i + j].getAddress() == addr && bundle[i + j].getValue() == 0) {
```

Comment:
Are you sure that dropping the loop and using indexing has the same effect?
That means the bundle should already be filled with empty transactions for signatures, is that the case? If so, is it order the same as what seem to be used in JS library?